### PR TITLE
Perf optimization: GetPodQOS() returns persisted value of PodStatus.QOSClass, if set.

### PR DESCRIPTION
**Pull request by vinaykul**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Invoking GetPodQOS is much more heavyweight op compared to reading PodStatus.QOSClass, a value that does not change after initial assignment because QoS class of pods are immutable. Running in-place pod resize e2e tests currently has GetPodQOS showing up in only 0.1% in the samples of kubelet flamegraph with 100 other pods on the node, each pod with 3 Burstable QoS class containers (real world cost may be higher or lower depending on nature of workloads) The intent is to eliminate that 0.1% and keep it that way by (re)setting this precedent in light of [KEP 2527](https://github.com/kubernetes/enhancements/pull/2537) allowing to use status as reliable, persistent source of truth.

GetPodQOS

#### Which issue(s) this PR fixes:

Fixes #
partially fixes 109547

#### Special notes for your reviewer:
In the flamegraph attached below, please see the sampling percentage of SetPodResizeStatus for this e2e pod resize test scenario. If this small optimization trial balloon PR stands as a valid use of KEP 2527, we may also have the case for relying on PodStatus.Resize and PodStatus...AllocatedResources as the reliable and persistent source of truth instead of checkpoint file - a good optimization imho (with special handling for stand-alone kubelet in case we want to support in-place resize for that case)

#### Does this PR introduce a user-facing change?

```release-note
GetPodQOS(pod *core.Pod) function now returns the stored value from PodStatus.QOSClass, if set. To compute/evaluate the value of QOSClass from scratch, ComputePodQOS(pod *core.Pod) must be used.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources
```

Kubelet flamegraph data:
kubelet-getpodqos

### CODE REVIEW

File: pkg/apis/core/helper/qos/qos.go [hunk 1]
```diff
@@ -30,6 +30,15 @@ func isSupportedQoSComputeResource(name core.ResourceName) bool {
 	return supportedQoSComputeResources.Has(string(name))
 }
 
+// PodQOSClass returns the QoS class of a pod persisted in the PodStatus.
+// If QOSClass is empty, it returns value of GetPodQOS() which computes pod's QoS class.
+func PodQOSClass(pod *core.Pod) core.PodQOSClass {
+	if pod.Status.QOSClass != "" {
+		return pod.Status.QOSClass
+	}
+	return GetPodQOS(pod)
```
Comment (thockin) [inline pkg/apis/core/helper/qos/qos.go:39]: Maybe note that this is much more expensive?

File: pkg/apis/core/v1/helper/qos/qos.go [hunk 2]
```diff
@@ -32,6 +32,15 @@ func isSupportedQoSComputeResource(name v1.ResourceName) bool {
 	return supportedQoSComputeResources.Has(string(name))
 }
 
+// PodQOSClass returns the QoS class of a pod persisted in the PodStatus.
+// If QOSClass is empty, it returns value of GetPodQOS() which computes pod's QoS class.
+func PodQOSClass(pod *v1.Pod) v1.PodQOSClass {
+	if pod.Status.QOSClass != "" {
+		return pod.Status.QOSClass
+	}
+	return GetPodQOS(pod)
```
Comment (thockin) [inline pkg/apis/core/v1/helper/qos/qos.go:41]: same

File: pkg/apis/core/helper/qos/qos.go (hunk 1 above)
Comment (thockin) [reply pkg/apis/core/helper/qos/qos.go:39]: Or even rename to GetPodQOSExpensive() or something?

Comment (thockin) [reply pkg/apis/core/helper/qos/qos.go:39]: Or name this  `GetPodQOS()` and move the old, expensive one to `getPodQOSExpensive()`

Comment (vinaykul) [reply pkg/apis/core/helper/qos/qos.go:39]: That's a great suggestion, fewer changes! I'll rename the current expensive one as ComputePodQOS() or DeterminePodQOS(), I'll update & retest over the weekend.

**Review by iholder101 — COMMENTED** [review]
LGTM! Thanks @vinaykul!
see one nit below

**Comment by iholder101** [comment]
/lgtm
Thank you @vinaykul!

File: pkg/apis/core/helper/qos/qos.go (hunk 1 above)
Comment (vinaykul) [reply pkg/apis/core/helper/qos/qos.go]: @thockin My last change has addressed this. PTAL, ty

**Comment by vinaykul** [comment]
@thockin PTAL

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Pull request merged**